🌊 Streams: Forkable classic streams

flash1293 opened the pull request:
Makes it possible to fork classic streams following the DSNS and have wired substreams for them.

Each DSNS classic stream behaves like the root `logs` - it's possible to create children via `_fork` or via `PUT` ing a child stream:
```
POST /api/streams/logs-mylogs-default/_fork
{
  "stream": { "name": "logs-mylogs.wiredchild-default" },
  "condition": { ... }
}

PUT /api/streams/logs-mylogs.wiredchild-default
{
  "ingest": {
    "routing": [],
    "processing": [],
    "wired": {
       "fields": {}
    }
  }
}
```

The children are wired and behave exactly like wired children of `logs` (inheriting fields and so on).

Wired streams report which root they belong to as part of the GET API (this can be leveraged in the UI):
```
GET kbn:/api/streams/logs-mylogs.wiredchild-default
{
  "name": "logs-mylogs.wiredchild-default",
  "stream": {
    "ingest": {
      "processing": [],
      "routing": [],
      "wired": {
        "fields": {}
      }
    }
  },
  "dashboards": [],
  "lifecycle": {
    "type": "dlm"
  },
  "root": "logs-mylogs-default",
  "inherited_fields": {}
}
```

Differences:
* The wired children of a classic stream do not inherit the default fields `logs` has - instead it starts completely empty except for the `@timestamp` field.
* The default processing of the `logs` stream (e.g. parsing JSON in the message field) does not happen

This PR does not adjust the UI yet, this will happen in a separate PR.


## Code guide

### kbn-streams-schema/src/helpers

We parse and created stream names all over the place using `split('.')` and `join('.')`. As this doesn't always work the same anymore depending on which type of stream the user deals with, this PR introduces parse/stringify helpers for the different stream names so they can be edited in a more type-safe way. All the generation is pulled in there and the hierarchy helpers, which are based on the parse/stringify helpers.

### Lifecycle

If a classic data stream has a `.kibana_streams` representation, but the actual data stream is deleted, the current 
retention can't be inferred. This PR is handling this explicitly by passing an "error" lifecycle to the consumer

### lib/streams/client.ts

A bunch of places that only dealt with wired streams before now also deal with classic streams. This required a bunch of changes throughout the client and its helpers. The most impactful one is the determination of the `unwired root` - for wired DSNS streams we can't infer from the name what the classic stream is they originated from, we need to actually check which streams exist. For example, if you are dealing with `logs-mytest.abc.xyz-default`, is the classic stream it's wired to `logs-mytest-default` or `logs-mytest.abc-default`? Both are possible and the actual data needs to be checked. The algorithm for this works like this:
* Check the `.kibana_streams` stream definitions based on the prefix (`logs-<dataset>`), whether there is a definition that matches the prefix of the wired stream
* If none is found, it's possible the classic stream never got persisted, but there is an actual data stream for that name. In this case, it tries to find an existing data stream and creates the `.kibana_streams` representation of the classic stream on the fly along with the child

This "unwired root" needs to be passed around to a few places (e.g. it's required to calculate the stack of component templates to apply)

elasticmachine commented:
<!-- buildkite-pr-bot-trigger-comment -->:robot: Jobs for this PR can be triggered through checkboxes. :construction: 

:information_source: To trigger the CI, please tick the checkbox below :point_down: 

- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-pull-request --> Click to trigger **kibana-pull-request** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-project-from-pr --> Click to trigger **kibana-deploy-project-from-pr** for this PR!

elasticmachine commented:
<!-- buildkite-pr-bot-trigger-comment -->:robot: Jobs for this PR can be triggered through checkboxes. :construction: 

:information_source: To trigger the CI, please tick the checkbox below :point_down: 

- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-pull-request --> Click to trigger **kibana-pull-request** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-project-from-pr --> Click to trigger **kibana-deploy-project-from-pr** for this PR!

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/268909)
* Commit: 4fa4a1db891813415ddc91c392a993a4b1c3837a

### Failed CI Steps
* [Jest Tests #8](https://buildkite.com/elastic/kibana-pull-request/builds/268909#01948998-0fb9-4228-af0d-c02d4d335b11)
* [Jest Tests #8](https://buildkite.com/elastic/kibana-pull-request/builds/268909#019489bb-5d69-4bd2-87a9-8ff3982d6ee5)
* [Jest Tests #15](https://buildkite.com/elastic/kibana-pull-request/builds/268909#01948998-0fba-4870-a726-bd961790081c)
* [FTR Configs #114](https://buildkite.com/elastic/kibana-pull-request/builds/268909#01948998-1024-4c88-bba5-5e01b8434f4d)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/268909#01948998-0fb9-4228-af0d-c02d4d335b11) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/268909/jobs/01948998-0fb9-4228-af0d-c02d4d335b11/artifacts/019489bb-4a16-439e-b2d6-e20168f21aac) Jest Tests #<span></span>8 / ReadStreamResponse should successfully parse
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/268909#019489bb-5d69-4bd2-87a9-8ff3982d6ee5) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/268909/jobs/019489bb-5d69-4bd2-87a9-8ff3982d6ee5/artifacts/019489bf-b392-4e35-ae21-b43dbae12de5) Jest Tests #<span></span>8 / ReadStreamResponse should successfully parse
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/268909#01948998-1024-4c88-bba5-5e01b8434f4d) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/268909/jobs/01948998-1024-4c88-bba5-5e01b8434f4d/artifacts/019489b8-f434-45b3-b73b-d57d3ef90252) FTR Configs #<span></span>114 / serverless security UI navigation navigate using search
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/268909#01948998-0fba-4870-a726-bd961790081c) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/268909/jobs/01948998-0fba-4870-a726-bd961790081c/artifacts/019489be-0d65-4b8a-9f64-b77d491391fb) Jest Tests #<span></span>15 / TemplatesList calls onDeleteTemplate correctly

### History
* :broken_heart: [Build #268704](https://buildkite.com/elastic/kibana-pull-request/builds/268704) failed 6a3ef017c07e2d79e0a8d998923d495f7cac2b40

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/268909","number":268909,"commit":"4fa4a1db891813415ddc91c392a993a4b1c3837a"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/268704","number":268704,"commit":"6a3ef017c07e2d79e0a8d998923d495f7cac2b40"}],"number":268909}
buildkite-pr-comment-->

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/269118)
* Commit: 2f3e1242d5a4b03d3f60e6897d64a7565b08a6e8

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/b0b44bbd22f7be4e651f4003bea7f3adc5c9f2a8) | [after](https://github.com/elastic/kibana/commit/2f3e1242d5a4b03d3f60e6897d64a7565b08a6e8) | diff |
| --- | --- | --- | --- |
| `streamsApp` | 273 | 277 | +4 |

#### Public APIs missing comments
> Total count of every public API that lacks a comment. Target amount is 0. Run `node scripts/build_api_docs --plugin [yourplugin] --stats comments` for more detailed information.


| id | [before](https://github.com/elastic/kibana/commit/b0b44bbd22f7be4e651f4003bea7f3adc5c9f2a8) | [after](https://github.com/elastic/kibana/commit/2f3e1242d5a4b03d3f60e6897d64a7565b08a6e8) | diff |
| --- | --- | --- | --- |
| `@kbn/streams-schema` | 109 | 130 | +21 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/b0b44bbd22f7be4e651f4003bea7f3adc5c9f2a8) | [after](https://github.com/elastic/kibana/commit/2f3e1242d5a4b03d3f60e6897d64a7565b08a6e8) | diff |
| --- | --- | --- | --- |
| `streamsApp` | 273.1KB | 273.8KB | +678.0B |

<details><summary>Unknown metric groups</summary>

#### API count
| id | [before](https://github.com/elastic/kibana/commit/b0b44bbd22f7be4e651f4003bea7f3adc5c9f2a8) | [after](https://github.com/elastic/kibana/commit/2f3e1242d5a4b03d3f60e6897d64a7565b08a6e8) | diff |
| --- | --- | --- | --- |
| `@kbn/streams-schema` | 109 | 130 | +21 |


</details>


### History
* :broken_heart: [Build #268909](https://buildkite.com/elastic/kibana-pull-request/builds/268909) failed 4fa4a1db891813415ddc91c392a993a4b1c3837a
* :broken_heart: [Build #268704](https://buildkite.com/elastic/kibana-pull-request/builds/268704) failed 6a3ef017c07e2d79e0a8d998923d495f7cac2b40

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/269118","number":269118,"commit":"2f3e1242d5a4b03d3f60e6897d64a7565b08a6e8"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/268909","number":268909,"commit":"4fa4a1db891813415ddc91c392a993a4b1c3837a"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/268704","number":268704,"commit":"6a3ef017c07e2d79e0a8d998923d495f7cac2b40"}],"number":269118}
buildkite-pr-comment-->

miltonhultgren commented on x-pack/solutions/observability/plugins/streams/server/lib/streams/client.ts:
```diff
@@ -739,38 +791,124 @@ export class StreamsClient {
+        .filter((candidateStream) => !isWiredStream(candidateStream))
+        .forEach((candidateStream) => {
+          const candidateStreamName = parseStreamName(candidateStream.name);
+          // The longest unwired stream with a matching prefix is the classic root
```
What happens if you have two classic streams like:
```
logs-mydata
logs-mydata.subset
```

And then fork both of them to some wired children? is that allowed? 
I _think_ this code will try to look for `logs-mydata` as the prefix in both cases and then for both cases return the longer stream name?

flash1293 replied:
it will return the longest that matches. So you should be able to have classic streams `logs-mydata-default` and `logs-mydata.subset-default`, then fork both of them into `logs-mydata.myfork-default` and `logs-mydata.subset.myfork-default` respectively.

I didn't test what happens when you try to fork `logs-mydata.subset-default` from `logs-mydata-default` even though both are classic streams already, that should be fixed as part of https://github.com/elastic/streams-program/issues/26

flash1293 replied:
However, it's a good point, I will add a test for this case

klacabane replied:
> I didn't test what happens when you try to fork logs-mydata.subset-default from logs-mydata-default even though both are classic streams already

You get this legitimate error `Cannot change unwired stream to wired stream`

miltonhultgren commented on x-pack/solutions/observability/plugins/streams/server/lib/streams/client.ts:
```diff
@@ -739,38 +791,124 @@ export class StreamsClient {
+    const streamName = parseStreamName(definition.name);
+    if (isWiredStream(definition) && streamName.type === 'dsns' && ancestors.length === 0) {
+      // We have a DSNS stream, which means it is not connected to the root logs stream, but no ancestors.
+      // This can happen if the classic stream for the unwired root hasn't been created yet.
```
```suggestion
      // This can happen if the classic stream definition for the unwired root hasn't been created yet.
```

Right? Because the (data) stream exists.

flash1293 replied:
right, should have been more exact

flash1293 commented on x-pack/solutions/observability/packages/kbn-streams-schema/src/helpers/hierarchy.ts:
```diff
@@ -5,32 +5,122 @@
+  if (parentStreamName.type === 'unknown' || childStreamName.type === 'unknown') {
+    return false;
+  }
+  if (parentStreamName.type === 'dsns' && childStreamName.type === 'dsns') {
```
It's either
* Wired to the root `logs` (and not DSNS)
* DSNS
* Unknown (everything else)

We decided a while back that streams should follow a new naming scheme to make conflicts less likely and make it easy for users to tell what they are dealing with: https://github.com/elastic/streams-program/blob/main/docs/design-decisions/2024-12-17-integration-compatibility.md

The wired children of classic streams do water that down a bit again, but I think of that mostly in the context of a migration use case.

miltonhultgren replied:
So a wired stream cannot define a namespace? Or is that just another segment in the dataset?

flash1293 commented on x-pack/solutions/observability/packages/kbn-streams-schema/src/helpers/hierarchy.ts:
```diff
@@ -5,32 +5,122 @@
-export function getAncestors(id: string) {
-  const parts = id.split('.');
-  return parts.slice(0, parts.length - 1).map((_, index) => parts.slice(0, index + 1).join('.'));
+export function getAncestors(id: string, unwiredRootId?: string) {
```
dsnsRootId makes sense to me, will update.

flash1293 replied:
actually, the concept is that of an unwired root, it just happens to always be a dsns so far, but that might change. The function can be extended to handle other types of unwired roots later on if we decide so.

I'll keep unwired root for now

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/269319)
* Commit: f6223b31a25d35787a3183643b323b307f0bcdff

### Failed CI Steps
* [Jest Tests #19](https://buildkite.com/elastic/kibana-pull-request/builds/269319#01948ea1-7ab3-4985-aba8-c7c7a2938fc7)
* [Jest Tests #19](https://buildkite.com/elastic/kibana-pull-request/builds/269319#01948ed0-1d0a-4cf6-be9f-ac2001745931)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/269319#01948ea1-7ab3-4985-aba8-c7c7a2938fc7) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/269319/jobs/01948ea1-7ab3-4985-aba8-c7c7a2938fc7/artifacts/01948ecf-f6f9-461c-806e-ede6fae29eba) Jest Tests #<span></span>19 / useGetActionLicense unhappy path
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/269319#01948ed0-1d0a-4cf6-be9f-ac2001745931) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/269319/jobs/01948ed0-1d0a-4cf6-be9f-ac2001745931/artifacts/01948efe-17f9-4578-8b16-1ad93da52aa3) Jest Tests #<span></span>19 / useGetCasesMetrics shows a toast error when the api return an error

### History
* :green_heart: [Build #269118](https://buildkite.com/elastic/kibana-pull-request/builds/269118) succeeded 2f3e1242d5a4b03d3f60e6897d64a7565b08a6e8
* :broken_heart: [Build #268909](https://buildkite.com/elastic/kibana-pull-request/builds/268909) failed 4fa4a1db891813415ddc91c392a993a4b1c3837a
* :broken_heart: [Build #268704](https://buildkite.com/elastic/kibana-pull-request/builds/268704) failed 6a3ef017c07e2d79e0a8d998923d495f7cac2b40

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/269319","number":269319,"commit":"f6223b31a25d35787a3183643b323b307f0bcdff"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/269118","number":269118,"commit":"2f3e1242d5a4b03d3f60e6897d64a7565b08a6e8"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/268909","number":268909,"commit":"4fa4a1db891813415ddc91c392a993a4b1c3837a"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/268704","number":268704,"commit":"6a3ef017c07e2d79e0a8d998923d495f7cac2b40"}],"number":269319}
buildkite-pr-comment-->

dgieselaar commented on x-pack/solutions/observability/plugins/streams/server/lib/streams/client.ts:
```diff
@@ -179,14 +189,19 @@ export class StreamsClient {
     } else if (isIngestStream(definition)) {
+      const dataStream = await this.getDataStream(definition.name);
+      if (!dataStream) {
+        throw new MalformedStream(`Data stream ${definition.name} could not be found`);
```
this should already throw a 404 if it's not found, right?

flash1293 replied:
That's mostly to make the type checker happy - better than the `!` operator and a weird arcane error a couple months and refactorings down the line.

dgieselaar replied:
yeah but shouldn't getDataStream itself throw a 404 if it isn't found?

flash1293 replied:
No, in some paths this is a valid case we should handle - if for example a user adds a dashboard and some processing to a classic stream and then the underlying data stream is deleted, it should still show and indicate that it's "orphaned". This is not fully implemented today, but I will complete it in a follow-up.

In this particular place as you mentioned we can't really complete the request, so it throws here.

dgieselaar replied:
What i mean is, any GET call with a specific id imho should fail with a 404, and not a 200 and undefined - which is to say, i think getDataStream() should return the data stream or return a 404 if not found, and then handling that error is up to the consumer. If we allow it to return undefined we end up pushing the assertion to the user which i think we will want to avoid

flash1293 replied:
> any GET call with a specific id imho should fail with a 404, and not a 200 and undefined

For sure, on the API level that is how it works. but the internal `getDataStream` method is one level below it and has different semantics.

> If we allow it to return undefined we end up pushing the assertion to the user which i think we will want to avoid

Why do you think so? I think it's a sensible API with clear semantics and we enforce via types that the consumer does the right thing. The alternative would be to have a throwing `getDataStream` and an `existsDataStream`, but that would imply two roundtrips we can avoid here and I think in most places we want to handle the situation that a data stream we thought would exist in a graceful way instead of throwing. Like here: https://github.com/elastic/kibana/blob/8d4a70c5e5a0ad9fd985d71d2f107bc29b330a92/x-pack/solutions/observability/plugins/streams/server/routes/streams/crud/read_stream.ts#L49

flash1293 commented on x-pack/solutions/observability/plugins/streams/server/lib/streams/client.ts:
```diff
@@ -617,15 +649,30 @@ export class StreamsClient {
+    const { logger, scopedClusterClient } = this.dependencies;
 
     if (!isWiredStream(definition)) {
+      for (const child of definition.stream.ingest.routing) {
```
refactored deleteStreamFromDefinition a bit to make it clearer.

flash1293 commented on x-pack/solutions/observability/packages/kbn-streams-schema/src/helpers/hierarchy.ts:
```diff
@@ -5,32 +5,122 @@
+  return false;
+}
+
+export function isRoot(name: string) {
```
urgh, refactoring foo

elasticmachine commented:
## :broken_heart: Build Failed
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/269435)
* Commit: 78b34db16a484ee11746751f261117ebd673beac

### Failed CI Steps
* [Check Types](https://buildkite.com/elastic/kibana-pull-request/builds/269435#01948f89-07fd-4ac5-8104-1b675548f8c5)

### History
* :broken_heart: [Build #269319](https://buildkite.com/elastic/kibana-pull-request/builds/269319) failed f6223b31a25d35787a3183643b323b307f0bcdff
* :green_heart: [Build #269118](https://buildkite.com/elastic/kibana-pull-request/builds/269118) succeeded 2f3e1242d5a4b03d3f60e6897d64a7565b08a6e8
* :broken_heart: [Build #268909](https://buildkite.com/elastic/kibana-pull-request/builds/268909) failed 4fa4a1db891813415ddc91c392a993a4b1c3837a
* :broken_heart: [Build #268704](https://buildkite.com/elastic/kibana-pull-request/builds/268704) failed 6a3ef017c07e2d79e0a8d998923d495f7cac2b40

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/269435","number":269435,"commit":"78b34db16a484ee11746751f261117ebd673beac"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/269319","number":269319,"commit":"f6223b31a25d35787a3183643b323b307f0bcdff"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/269118","number":269118,"commit":"2f3e1242d5a4b03d3f60e6897d64a7565b08a6e8"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/268909","number":268909,"commit":"4fa4a1db891813415ddc91c392a993a4b1c3837a"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/268704","number":268704,"commit":"6a3ef017c07e2d79e0a8d998923d495f7cac2b40"}],"number":269435}
buildkite-pr-comment-->

klacabane commented:
Can we adjust the `MalformedStreamId` error messages to account for dsns, we currently get this error which is a bit confusing
```
{
  "statusCode": 400,
  "error": "Bad Request",
  "message": "The ID (foo-dataset-namespace.wired) from the new stream must start with the parent's id (foo-dataset-namespace), followed by a dot and a name",
  "attributes": {
    "data": null
  }
}
```

klacabane commented:
Logging errors I get while testing this, didn't really look at the code yet

_1. This one probably needs a better error message (eg a stream with that name already exists)_
```
PUT _data_stream/foo-namespace-default
PUT _data_stream/foo-namespace.foo-default

POST kbn:/api/streams/foo-namespace-default/_fork
{
    "stream": { "name": "foo-namespace.foo-default" },
    "if": { "never": {} }
}

--->
{
  "statusCode": 400,
  "error": "Bad Request",
  "message": "Cannot change unwired stream to wired stream",
  "attributes": {
    "data": null
  }
}
```

2. I expected this one to work and pick `foo-namespace.foo-default` as root
```
PUT _data_stream/foo-namespace-default
PUT _data_stream/foo-namespace.foo-default

PUT kbn:/api/streams/foo-namespace.foo.bar-default
{
    "stream": {
        "ingest": {
            "processing": [],
            "routing": [],
            "wired": { "fields": {} }
        }
    },
    "dashboards": []
}

--->
{
  "statusCode": 400,
  "error": "Bad Request",
  "message": "Cannot change unwired stream to wired stream",
  "attributes": {
    "data": null
  }
}
```

klacabane commented:
I expected this one to work and pick `foo-namespace.foo-default` as root but it failed. Looks like it picks `foo-namespace-default` as root and tries to create intermediate streams but they already exist are unwired stream
```
PUT _data_stream/foo-namespace-default
PUT _data_stream/foo-namespace.foo-default

PUT kbn:/api/streams/foo-namespace.foo.bar-default
{
    "stream": {
        "ingest": {
            "processing": [],
            "routing": [],
            "wired": { "fields": {} }
        }
    },
    "dashboards": []
}

--->
{
  "statusCode": 400,
  "error": "Bad Request",
  "message": "Cannot change unwired stream to wired stream",
  "attributes": {
    "data": null
  }
}
```

flash1293 commented:
> Can we adjust the MalformedStreamId error messages to account for dsns, we currently get this error which is a bit confusing

Done 👍

flash1293 commented:
> I expected this one to work and pick foo-namespace.foo-default as root but it failed. Looks like it picks foo-namespace-default as root and tries to create intermediate streams but they already exist as unwired stream

I expected this to work too, definitely a bug - will look into it and add a test case.

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/269530)
* Commit: 3c3fb27470d4694ce803c3687e1114be538adb1f

### Failed CI Steps
* [Jest Tests #2](https://buildkite.com/elastic/kibana-pull-request/builds/269530#0194923b-fdb5-4ead-ad41-525d4984a590)
* [Jest Tests #2](https://buildkite.com/elastic/kibana-pull-request/builds/269530#0194925e-0380-4c12-9d68-b716643732c6)

### History
* :broken_heart: [Build #269435](https://buildkite.com/elastic/kibana-pull-request/builds/269435) failed 78b34db16a484ee11746751f261117ebd673beac
* :broken_heart: [Build #269319](https://buildkite.com/elastic/kibana-pull-request/builds/269319) failed f6223b31a25d35787a3183643b323b307f0bcdff
* :green_heart: [Build #269118](https://buildkite.com/elastic/kibana-pull-request/builds/269118) succeeded 2f3e1242d5a4b03d3f60e6897d64a7565b08a6e8
* :broken_heart: [Build #268909](https://buildkite.com/elastic/kibana-pull-request/builds/268909) failed 4fa4a1db891813415ddc91c392a993a4b1c3837a
* :broken_heart: [Build #268704](https://buildkite.com/elastic/kibana-pull-request/builds/268704) failed 6a3ef017c07e2d79e0a8d998923d495f7cac2b40

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/269530","number":269530,"commit":"3c3fb27470d4694ce803c3687e1114be538adb1f"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/269435","number":269435,"commit":"78b34db16a484ee11746751f261117ebd673beac"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/269319","number":269319,"commit":"f6223b31a25d35787a3183643b323b307f0bcdff"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/269118","number":269118,"commit":"2f3e1242d5a4b03d3f60e6897d64a7565b08a6e8"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/268909","number":268909,"commit":"4fa4a1db891813415ddc91c392a993a4b1c3837a"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/268704","number":268704,"commit":"6a3ef017c07e2d79e0a8d998923d495f7cac2b40"}],"number":269530}
buildkite-pr-comment-->

flash1293 commented:
@klacabane should be ready for another look - the problem was that it didn't pick the longest match when the streams were never saved before. I refactored a bit to make it clearer (the whole logic is now in `getUnwiredRootForStream`)  and added an integration test for the case.

miltonhultgren commented on x-pack/test/api_integration/deployment_agnostic/apis/observability/streams/classic_forking.ts:
```diff
@@ -325,5 +327,59 @@ export default function ({ getService }: DeploymentAgnosticFtrProviderContext) {
+      await esClient.indices.createDataStream({ name: 'logs-shared-default' });
+      await esClient.indices.createDataStream({ name: 'logs-shared.prefix-default' });
+
+      await putStream(apiClient, 'logs-shared.prefix.myfork-default', {
```
What happens to `logs-shared.prefix-default` if we first fork on `logs-shared-default` instead?

flash1293 replied:
It just sits there as an unwired stream, e.g. you could have unwired `logs-shared-default` with a wired child `logs-shared.myfork-default` and unwired `logs-shared.prefix-default` with a wired child `logs-shared.prefix.myfork-default`.

If a fork would cause a conflict, it responds a 400.

It's a good case to validate though, I added an integration test for it.

miltonhultgren commented on x-pack/solutions/observability/plugins/streams/server/lib/streams/client.ts:
```diff
@@ -430,6 +430,21 @@ export class StreamsClient {
     }
   }
 
+  private validateIsChildOf(parent: string, destination: string) {
```
Nit/question: Do we make any distinction between validating and **asserting** in the plugin? 

From the outside, this looks like an assertion (stop running further code if this condition is not met) compared to a validation (which has some outcome that is then acted on).

flash1293 replied:
Maybe assertion would be the better naming, no strong opinion.

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/269653)
* Commit: de2114e948afc9a54b0c21fe09333ce0d24773ca

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/5e8188a9602af21ab5815db4a739d6abb1b1218c) | [after](https://github.com/elastic/kibana/commit/de2114e948afc9a54b0c21fe09333ce0d24773ca) | diff |
| --- | --- | --- | --- |
| `streamsApp` | 282 | 286 | +4 |

#### Public APIs missing comments
> Total count of every public API that lacks a comment. Target amount is 0. Run `node scripts/build_api_docs --plugin [yourplugin] --stats comments` for more detailed information.


| id | [before](https://github.com/elastic/kibana/commit/5e8188a9602af21ab5815db4a739d6abb1b1218c) | [after](https://github.com/elastic/kibana/commit/de2114e948afc9a54b0c21fe09333ce0d24773ca) | diff |
| --- | --- | --- | --- |
| `@kbn/streams-schema` | 201 | 225 | +24 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/5e8188a9602af21ab5815db4a739d6abb1b1218c) | [after](https://github.com/elastic/kibana/commit/de2114e948afc9a54b0c21fe09333ce0d24773ca) | diff |
| --- | --- | --- | --- |
| `streamsApp` | 275.2KB | 275.9KB | +699.0B |

<details><summary>Unknown metric groups</summary>

#### API count
| id | [before](https://github.com/elastic/kibana/commit/5e8188a9602af21ab5815db4a739d6abb1b1218c) | [after](https://github.com/elastic/kibana/commit/de2114e948afc9a54b0c21fe09333ce0d24773ca) | diff |
| --- | --- | --- | --- |
| `@kbn/streams-schema` | 201 | 225 | +24 |


</details>


### History
* :broken_heart: [Build #269530](https://buildkite.com/elastic/kibana-pull-request/builds/269530) failed 3c3fb27470d4694ce803c3687e1114be538adb1f
* :broken_heart: [Build #269435](https://buildkite.com/elastic/kibana-pull-request/builds/269435) failed 78b34db16a484ee11746751f261117ebd673beac
* :broken_heart: [Build #269319](https://buildkite.com/elastic/kibana-pull-request/builds/269319) failed f6223b31a25d35787a3183643b323b307f0bcdff
* :green_heart: [Build #269118](https://buildkite.com/elastic/kibana-pull-request/builds/269118) succeeded 2f3e1242d5a4b03d3f60e6897d64a7565b08a6e8
* :broken_heart: [Build #268909](https://buildkite.com/elastic/kibana-pull-request/builds/268909) failed 4fa4a1db891813415ddc91c392a993a4b1c3837a

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/269653","number":269653,"commit":"de2114e948afc9a54b0c21fe09333ce0d24773ca"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/269530","number":269530,"commit":"3c3fb27470d4694ce803c3687e1114be538adb1f"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/269435","number":269435,"commit":"78b34db16a484ee11746751f261117ebd673beac"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/269319","number":269319,"commit":"f6223b31a25d35787a3183643b323b307f0bcdff"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/269118","number":269118,"commit":"2f3e1242d5a4b03d3f60e6897d64a7565b08a6e8"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/268909","number":268909,"commit":"4fa4a1db891813415ddc91c392a993a4b1c3837a"}],"number":269653}
buildkite-pr-comment-->

flash1293 commented:
Closing for now - we decided to keep the wired and unwired "worlds" separate to avoid user confusion.
